### PR TITLE
Closes #5 - Metrics on a per-process basis

### LINKED ISSUE #5: I want the ability to get process information for running nio services.
The current metrics block does not get info for individual processes. 

I want a block that can be passed a signal with a process id and then get an output signal with info about the process. That way you can build a service that first uses the nio api to get a process id and then pass that to his block to get memory info. 

Another option is to build a block that does that all in one. You can tell it what services you want to check (or all services that are not stopped) and emit process information at a configured interval.

### PR DESCRIPTION
**Pull request by oleiman**
Signed-off-by: Oren Leiman <oleiman@neutral.io>

### CODE REVIEW

File: process_metrics_block.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                                  attr_default=NoPIDException)
+
+    def __init__(self):
+        super().__init__()
+        self._retry_count = 0
+
+    def process_signals(self, signals):
+        for sig in signals:
+            try:
+                pid = self.pid_expr(sig)
```
Comment (hansmosh) [inline process_metrics_block.py]: If I remember correctly, expression properties can raise exceptions if the expression can't evaluate so these always need to be wrapped in a try (which you do) with an except clause to handle Exception.

File: process_metrics_block.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                stats = self._collect_stats(pid)
+                if stats:
+                    self.notify_signals([Signal(stats)])
```
Comment (hansmosh) [inline process_metrics_block.py]: For most blocks, we try to stick to the idea that we wait to notify all signals at the end of the iterating. If a list of signals come in together, then they should leave together.

File: process_metrics_block.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+            if self.menu.cmd_line:
+                result['cmd_line'] = ' '.join(proc.cmdline())
+                
+        except Exception as e:
+            self._logger.error(
+                "While processing system metrics: {0}".format(str(e))
+            )
+            if self._retry_count < RETRY_LIMIT:
+                self._retry_count += 1
```
Comment (hansmosh) [inline process_metrics_block.py]: This either needs to be reset to 0 on a success or it should be a local variable instead of self._retry_count

Comment (oleiman) [inline process_metrics_block.py]: you're absolutely right. this should also get changed in the basic metrics block...

File: process_metrics_block.py (hunk 2 above)
Comment (oleiman) [inline process_metrics_block.py]: fair enough

File: process_metrics_block.py (hunk 1 above)
Comment (oleiman) [inline process_metrics_block.py]: indeed

File: process_metrics_block.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                # if there's no PID in the signal, skip it
+                self._logger.debug(
+                    "Skipping signal {}: No PID".format(sig.to_dict()))
+                
+            except Exception as e:
+                self._logger.error(
+                    "Error while processing signal: {}: {}".format(
+                        type(e).__name__, str(e)))
+
+        self.notify_signals(results)
```
Comment (hansmosh) [inline process_metrics_block.py]: First test to make sure results is not empty.

**Pull request merged**